packages/go_router/test/delegate_test.dart

```diff
@@ -355,6 +355,27 @@ void main() {
         );
       },
     );
+
+    testWidgets('It should preserver source completer',
+        (WidgetTester tester) async {
+      final GoRouter goRouter = await createGoRouter(tester);
+
+      goRouter.push('/page-0');
+      await tester.pumpAndSettle();
+
+      final Future<Object?> sourceFeature = goRouter.push('/page-1');
+      await tester.pumpAndSettle();
+
+      goRouter.pushReplacement('/page-2');
+      await tester.pumpAndSettle();
+
+      expect(goRouter.routerDelegate.currentConfiguration.matches.length, 3);
+      goRouter.routerDelegate.pop(true);
+      await tester.pumpAndSettle();
+
+      expect(goRouter.routerDelegate.currentConfiguration.matches.length, 2);
+      expect(await sourceFeature, true);
```

Comment:
This does not look right. the once `page-1` is replaced, its future should never be completed. The `pop` result on line 373 is only meant for `page-2`. It should not be used to complete `page-1`.

Reply:
> This does not look right. the once `page-1` is replaced, its future should never be completed. The `pop` result on line 373 is only meant for `page-2`. It should not be used to complete `page-1`.

On the other hand, is it good approach to throw any Future into the void (in any case, not only in this particular case)?

This implies that users of go router must carefully design their flow, because feature never ends. And it differs from how the navigator works itself - it completes the future on replacement, which is actually pretty logically.

Reply:
> On the other hand, is it good approach to throw any Future into the void (in any case, not only in this particular case)?

No, but sometimes it is unavoidable.

> it differs from how the navigator works itself - it completes the future on replacement

Ah you are right, I missed that the navigator completes the previous route for pushReplacement, although the way it completes previous route is different from this pr. We can add a result parameter to pushReplacement to complete previous route. The current implementation uses the same result for both routes seem wrong to me.

The NavigatorState.replace however removes the route without completing it. I think we should keep it that way in go_router